File: Lib.Redist/Sources/Lexer/PrefetchedText.cs
/**********************************************************************
* Copyright (c) 2013 Laurent Wouters and others
* This program is free software: you can redistribute it and/or modify
* it under the terms of the GNU Lesser General Public License as
* published by the Free Software Foundation, either version 3
* of the License, or (at your option) any later version.
*
* This program is distributed in the hope that it will be useful,
* but WITHOUT ANY WARRANTY; without even the implied warranty of
* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
* GNU Lesser General Public License for more details.
*
* You should have received a copy of the GNU Lesser General
* Public License along with this program.
* If not, see <http://www.gnu.org/licenses/>.
*
* Contributors:
*     Laurent Wouters - lwouters@xowl.org
**********************************************************************/
using System.Collections.Generic;

namespace Hime.Redist.Lexer
{
	/// <summary>
	/// Stores the full content of an input lexer
	/// </summary>
	/// <remarks>
	/// All line numbers and column numbers are 1-based.
	/// Indices in the content are 0-based.
	/// </remarks>
	class PrefetchedText : TokenizedText
	{
		/// <summary>
		/// The initiaal size of the cache of line start indices
		/// </summary>
		private const int initLineCount = 10000;

		/// <summary>
		/// Represents the metadata of a token
		/// </summary>
		private struct Cell
		{
			/// <summary>
			/// The terminal's index
			/// </summary>
			public int terminal;

			/// <summary>
			/// Start index of the text
			/// </summary>
			public int start;

			/// <summary>
			/// Length of the token
			/// </summary>
			public int length;

			/// <summary>
			/// Initializes this cell
			/// </summary>
			/// <param name="terminal">The terminal's index</param>
			/// <param name="start">Start index of the text</param>
			/// <param name="length">Length of the token</param>
			public Cell(int terminal, int start, int length)
			{
				this.terminal = terminal;
				this.start = start;
				this.length = length;
			}
		}

		/// <summary>
		/// The full content of the input
		/// </summary>
		private string content;
		/// <summary>
		/// Cache of the starting indices of each line within the text
		/// </summary>
		private int[] lines;
		/// <summary>
		/// Index of the next line
		/// </summary>
		private int line;
		/// <summary>
		/// The terminal symbols matched in this content
		/// </summary>
		private IList<Symbol> terminals;
		/// <summary>
		/// The token data in this content
		/// </summary>
		private Utils.BigList<Cell> cells;

		/// <summary>
		/// Initializes this text
		/// </summary>
		/// <param name="terminals">The terminal symbols</param>
		/// <param name="content">The full lexer's input as a string</param>
		public PrefetchedText(IList<Symbol> terminals, string content)
		{
			this.content = content;
			this.terminals = terminals;
			this.cells = new Hime.Redist.Utils.BigList<Cell>();
		}

		#region Internal API
		/// <summary>
		/// Finds all the lines in this content
		/// </summary>
		public void FindLines()
		{
			this.lines = new int[initLineCount];
			this.line = 0;
			bool flagCR = false;
			for (int i = 0; i != content.Length; i++)
			{
				switch ((int)content[i])
				{
					case 0x0D:
						flagCR = true;
						AddLine(i);
						break;
					case 0x0A:
						if (!flagCR)
							AddLine(i);
						flagCR = false;
						break;
					case 0x0B:
					case 0x0C:
					case 0x85:
					case 0x2028:
					case 0x2029:
						flagCR = false;
						AddLine(i);
						break;
					default:
						flagCR = false;
						break;
				}
			}
			lines[line] = content.Length;
		}

		/// <summary>
		/// Adds a line starting at the specified index
		/// </summary>
		/// <param name="index">An index in the content</param>
		private void AddLine(int index)
		{
			if (line + 1 >= lines.Length)
			{
				int[] t = new int[lines.Length + initLineCount];
				System.Buffer.BlockCopy(lines, 0, t, 0, lines.Length * 4);
				lines = t;
			}
			lines[line++] = index;
		}

		/// <summary>
		/// Adds a detected token in this text
		/// </summary>
		/// <param name="terminal">Index of the matched terminal</param>
		/// <param name="start">Start index in the text</param>
		/// <param name="length">Length of the token</param>
		public void AddToken(int terminal, int start, int length)
		{
			cells.Add(new Cell(terminal, start, length));
		}

		/// <summary>
		/// Gets the token at the specified index
		/// </summary>
		/// <param name="index">A token's index</param>
		/// <returns>The token at the specified index</returns>
		public Token GetTokenAt(int index)
		{
			Cell cell = cells[index];
			return new Token(terminals[cell.terminal].ID, index);
		}
		#endregion


		#region Implementation of Text
		/// <summary>
		/// Gets the number of lines
		/// </summary>
		public int LineCount
		{
			get
			{
				if (lines == null)
					FindLines();
				return line;
			}
		}

		/// <summary>
		/// Gets the size in number of characters
		/// </summary>
		public int Size { get { return content.Length; } }

		/// <summary>
		/// Gets the substring beginning at the given index with the given length
		/// </summary>
		/// <param name="index">Index of the substring from the start</param>
		/// <param name="length">Length of the substring</param>
		/// <returns>The substring</returns>
		public string GetValue(int index, int length)
		{
			if (length == 0)
				return string.Empty;
			return content.Substring(index, length);
		}

		/// <summary>
		/// Gets the starting index of the i-th line
		/// </summary>
		/// <param name="line">The line number</param>
		/// <returns>The starting index of the line</returns>
		/// <remarks>The line numbering is 1-based</remarks>
		public int GetLineIndex(int line)
		{
			if (lines == null)
				FindLines();
			return lines[line - 1];
		}

		/// <summary>
		/// Gets the length of the i-th line
		/// </summary>
		/// <param name="line">The line number</param>
		/// <returns>The length of the line</returns>
		/// <remarks>The line numbering is 1-based</remarks>
		public int GetLineLength(int line)
		{
			if (lines == null)
				FindLines();
			return (lines[line] - lines[line - 1]);
		}

		/// <summary>
		/// Gets the string content of the i-th line
		/// </summary>
		/// <param name="line">The line number</param>
		/// <returns>The string content of the line</returns>
		/// <remarks>The line numbering is 1-based</remarks>
		public string GetLineContent(int line)
		{
			return GetValue(GetLineIndex(line), GetLineLength(line));
		}

		/// <summary>
		/// Gets the position at the given index
		/// </summary>
		/// <param name="index">Index from the start</param>
		/// <returns>The position (line and column) at the index</returns>
		public TextPosition GetPositionAt(int index)
		{
			int l = FindLineAt(index);
			return new TextPosition(l + 1, index - lines[l]);
		}

		/// <summary>
		/// Finds the 0-based number of the line at the given index in the content
		/// </summary>
		/// <param name="index">The index within this content</param>
		private int FindLineAt(int index)
		{
			if (lines == null)
				FindLines();
			int start = 0;
			int end = line - 1;
			while (true)
			{
				if (end == start || end == start + 1)
					return start;
				int m = (start + end) / 2;
				int v = lines[m];
				if (index == v)
					return m;
				if (index < v)
					end = m;
				else
					start = m;
			}
		}
		#endregion


		#region Implementation of IEnumerable
		/// <summary>
		/// Gets an enumerator of the contained tokens
		/// </summary>
		/// <returns>An enumerator of tokens</returns>
		public IEnumerator<Symbol> GetEnumerator()
		{
			return new SymbolEnumerator(this);
		}

		/// <summary>
		/// Gets an enumerator of the contained tokens
		/// </summary>
		/// <returns>An enumerator of tokens</returns>
		System.Collections.IEnumerator System.Collections.IEnumerable.GetEnumerator()
		{
			return new SymbolEnumerator(this);
		}

		/// <summary>
		/// Represents an enumerator of tokens
		/// </summary>
		private class SymbolEnumerator : IEnumerator<Symbol>
		{
			private PrefetchedText text;
			private int index;

			public SymbolEnumerator(PrefetchedText text)
			{
				this.text = text;
				this.index = -1;
			}

			public Symbol Current { get { return text[index]; } }

			object System.Collections.IEnumerator.Current { get { return text[index]; } }

			public bool MoveNext()
			{
				index++;
				return (index == text.cells.Size);
			}

			public void Reset()
			{
				index = -1;
			}

			public void Dispose()
			{
				text = null;
			}
		}
		#endregion


		#region Implementation of TokenizedText
		/// <summary>
		/// Gets the number of tokens in this text
		/// </summary>
		public int TokenCount { get { return cells.Size; } }

		/// <summary>
		/// Gets the token at the given index
		/// </summary>
		/// <param name="index">An index</param>
		/// <returns>The token</returns>
		public Symbol this[int index]
		{
			get
			{
				Cell cell = cells[index];
				Symbol terminal = terminals[cell.terminal];
				return new Symbol(terminal.ID, terminal.Name, GetValue(cell.start, cell.length));
			}
		}

		/// <summary>
		/// Gets the value of the token at the given index
		/// </summary>
		/// <param name="index">An index</param>
		/// <returns>The corresponding symbol</returns>
		public Symbol GetSymbolAt(int index)
		{
			Cell cell = cells[index];
			Symbol terminal = terminals[cell.terminal];
			string value = GetValue(cell.start, cell.length);
			return new Symbol(terminal.ID, terminal.Name, value);
		}

		/// <summary>
		/// Gets the position of the token at the given index
		/// </summary>
		/// <param name="tokenIndex">The index of a token</param>
		/// <returns>The position (line and column) of the token</returns>
		public TextPosition GetPositionOf(int tokenIndex)
		{
			Cell cell = cells[tokenIndex];
			return GetPositionAt(cell.start);
		}
		#endregion
	}
}

File: runtimes/java/src/org/xowl/hime/redist/lexer/PrefetchedText.java
/**********************************************************************
 * Copyright (c) 2014 Laurent Wouters and others
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as
 * published by the Free Software Foundation, either version 3
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General
 * Public License along with this program.
 * If not, see <http://www.gnu.org/licenses/>.
 *
 * Contributors:
 *     Laurent Wouters - lwouters@xowl.org
 **********************************************************************/

package org.xowl.hime.redist.lexer;

import org.xowl.hime.redist.*;
import org.xowl.hime.redist.utils.BigList;

import java.util.Arrays;
import java.util.List;

/**
 * Stores the full content of an input lexer
 *
 * All line numbers and column numbers are 1-based.
 * Indices in the content are 0-based.
 */
public class PrefetchedText implements TokenizedText {
    /**
     * The initiaal size of the cache of line start indices
     */
    private static final int initLineCount = 10000;

    /**
     * Represents the metadata of a token
     */
    private static class Cell {
        /**
         * The terminal's index
         */
        public int terminal;
        /**
         * Start index of the text
         */
        public int start;
        /**
         * Length of the token
         */
        public int length;

        /**
         * Initializes this cell
         *
         * @param terminal The terminal's index
         * @param start    Start index of the text
         * @param length   Length of the token
         */
        public Cell(int terminal, int start, int length) {
            this.terminal = terminal;
            this.start = start;
            this.length = length;
        }
    }

    /**
     * The full content of the input
     */
    private String content;
    /**
     * Cache of the starting indices of each line within the text
     */
    private int[] lines;
    /**
     * Index of the next line
     */
    private int line;
    /**
     * The terminal symbols matched in this content
     */
    private List<Symbol> terminals;
    /**
     * The token data in this content
     */
    private BigList<Cell> cells;

    /**
     * Initializes this text
     *
     * @param terminals The terminal symbols
     * @param content   The full lexer's input as a string
     */
    public PrefetchedText(List<Symbol> terminals, String content) {
        this.content = content;
        this.terminals = terminals;
        this.cells = new BigList<Cell>(Cell.class, Cell[].class);
    }

    /**
     * Finds all the lines in this content
     */
    public void findLines() {
        this.lines = new int[initLineCount];
        this.lines[0] = 0;
        this.line = 1;
        char c1 = '\0';
        char c2 = '\0';
        for (int i = 0; i != content.length(); i++) {
            c1 = c2;
            c2 = content.charAt(i);
            // is c1 c2 a line ending sequence?
            if (isLineEnding(c1, c2)) {
                // are we late to detect MacOS style?
                if (c1 == '\r' && c2 != '\n')
                    addLine(i);
                else
                    addLine(i + 1);
            }
        }
    }

    /**
     * Determines whether [c1, c2] form a line ending sequence
     *
     * Recognized sequences are:
     * [U+000D, U+000A] (this is Windows-style \r \n)
     * [U+????, U+000A] (this is unix style \n)
     * [U+000D, U+????] (this is MacOS style \r, without \n after)
     * Others:
     * [?, U+000B], [?, U+000C], [?, U+0085], [?, U+2028], [?, U+2029]
     *
     * @param c1 First character
     * @param c2 Second character
     * @return true  if this is a line ending sequence
     */
    private boolean isLineEnding(char c1, char c2) {
        // other characters
        if (c2 == '\u000B' || c2 == '\u000C' || c2 == '\u0085' || c2 == '\u2028' || c2 == '\u2029')
            return true;
        // matches [\r, \n] [\r, ??] and  [??, \n]
        if (c1 == '\r' || c2 == '\n')
            return true;
        return false;
    }

    /**
     * Adds a line starting at the specified index
     *
     * @param index An index in the content
     */
    private void addLine(int index) {
        if (line >= lines.length)
            lines = Arrays.copyOf(lines, lines.length + initLineCount);
        lines[line] = index;
        line++;
    }

    /**
     * Adds a detected token in this text
     *
     * @param terminal Index of the matched terminal
     * @param start    Start index in the text
     * @param length   Length of the token
     */
    public void addToken(int terminal, int start, int length) {
        cells.add(new Cell(terminal, start, length));
    }

    /**
     * Gets the token at the specified index
     *
     * @param index A token's index
     * @return The token at the specified index
     */
    public Token getTokenAt(int index) {
        Cell cell = cells.get(index);
        return new Token(terminals.get(cell.terminal).getID(), index);
    }


    /**
     * Gets the number of lines
     *
     * @return The number of lines
     */
    public int getLineCount() {
        if (lines == null)
            findLines();
        return line;
    }

    /**
     * Gets the size in number of characters
     *
     * @return The size in number of characters
     */
    public int size() {
        return content.length();
    }

    /**
     * Gets the substring beginning at the given index with the given length
     *
     * @param index  Index of the substring from the start
     * @param length Length of the substring
     * @return The substring
     */
    public String getValue(int index, int length) {
        if (length == 0)
            return "";
        return content.substring(index, index + length);
    }

    /**
     * Gets the starting index of the i-th line
     *
     * The line numbering is 1-based
     *
     * @param line The line number
     * @return The starting index of the line
     */
    public int getLineIndex(int line) {
        if (lines == null)
            findLines();
        return lines[line - 1];
    }

    /**
     * Gets the length of the i-th line
     *
     * The line numbering is 1-based
     *
     * @param line The line number
     * @return The length of the line
     */
    public int getLineLength(int line) {
        if (lines == null)
            findLines();
        if (line == this.line)
            return (content.length() - lines[this.line - 1]);
        return (lines[line] - lines[line - 1]);
    }

    /**
     * Gets the string content of the i-th line
     *
     * The line numbering is 1-based
     *
     * @param line The line number
     * @return The string content of the line
     */
    public String getLineContent(int line) {
        return getValue(getLineIndex(line), getLineLength(line));
    }

    /**
     * Gets the position at the given index
     *
     * @param index Index from the start
     * @return The position (line and column) at the index
     */
    public TextPosition getPositionAt(int index) {
        int l = findLineAt(index);
        return new TextPosition(l + 1, index - lines[l] + 1);
    }

    /**
     * Gets the context description for the current text at the specified position
     *
     * @param position The position in this text
     * @return The context description
     */
    public Context getContext(TextPosition position) {
        String content = getLineContent(position.getLine());
        if (content.length() == 0)
            return new Context("", "^");
        int end = content.length() - 1;
        while (end != -1 && (content.charAt(end) == '\n' || content.charAt(end) == '\r'))
            end--;
        int start = 0;
        while (start < end && Character.isWhitespace(content.charAt(start)))
            start++;
        if (position.getColumn() - 1 < start) {
            // the position is in the whitespace prefix ...
            start = 0;
        }
        if (position.getColumn() - 1 > end) {
            // the position is in the trailing line endings ...
            end = content.length() - 1;
        }
        StringBuilder builder = new StringBuilder();
        for (int i=start; i!=position.getColumn() - 1; i++)
            builder.append(content.charAt(i) == '\t' ? '\t' : ' ');
        builder.append("^");
        return new Context( content.substring(start, end + 1), builder.toString());
    }

    /**
     * Finds the index in the cache of the line at the given input index in the content
     *
     * @param index The index within this content
     * @return The index of the corresponding line in the cache
     */
    private int findLineAt(int index) {
        if (lines == null)
            findLines();
        for (int i=1; i!=line; i++) {
            if (index < lines[i]) {
                return i - 1;
            }
        }
        return line - 1;
    }

    /**
     * Gets the number of tokens in this text
     *
     * @return The number of tokens in this text
     */
    public int getTokenCount() {
        return cells.size();
    }

    /**
     * Gets the token at the given index
     *
     * @param index An index
     * @return The token
     */
    public Symbol at(int index) {
        Cell cell = cells.get(index);
        Symbol terminal = terminals.get(cell.terminal);
        String value = getValue(cell.start, cell.length);
        return new Symbol(terminal.getID(), terminal.getName(), value);
    }

    /**
     * Gets the position of the token at the given index
     *
     * @param tokenIndex The index of a token
     * @return The position (line and column) of the token
     */
    public TextPosition getPositionOf(int tokenIndex) {
        Cell cell = cells.get(tokenIndex);
        return getPositionAt(cell.start);
    }
}
